packages/next/src/shared/lib/router/utils/route-regex.ts

```diff
@@ -17,15 +17,51 @@ export interface RouteRegex {
   re: RegExp
 }
 
+/**
+ * Regular expression pattern used to match route parameters.
+ * Matches both single parameters and parameter groups.
+ * Examples:
+ *   - `[[...slug]]` matches parameter group with key 'slug', repeat: true, optional: true
+ *   - `[...slug]` matches parameter group with key 'slug', repeat: true, optional: false
+ *   - `[[foo]]` matches parameter with key 'foo', repeat: false, optional: true
+ *   - `[bar]` matches parameter with key 'bar', repeat: false, optional: false
+ */
+const PARAMETER_PATTERN = /\[((?:\[.*\])|.+)\]/
+
 /**
  * Parses a given parameter from a route to a data structure that can be used
- * to generate the parametrized route. Examples:
+ * to generate the parametrized route.
+ * Examples:
+ *   - `[[...slug]]` -> `{ key: 'slug', repeat: true, optional: true }`
+ *   - `[...slug]` -> `{ key: 'slug', repeat: true, optional: false }`
+ *   - `[[foo]]` -> `{ key: 'foo', repeat: false, optional: true }`
+ *   - `[bar]` -> `{ key: 'bar', repeat: false, optional: false }`
```

Comment:
Looking at https://github.com/vercel/next.js/pull/65277/files#diff-6d649b04b65094a03477d17ce26c376ebc0a7657ddbd2787ef93dd1348fb5892L29, seems that all of these cases that use dynamic params should be considered "optional" right?

Reply:
No, based on the regexp 

```js
const PARAMETER_PATTERN = /\[((?:\[.*\])|.+)\]/
```

We take the first match and pass that into `parseMatchedParameter` which in the example of `[[...foo]]` will be `[...foo]`.  Thats why `parseMatchedParameter` checks for `startsWith([)` and `endsWith(])` sets optional to `true`.  In the example of `[...foo]`, `parseMatchParameter` received `...foo` and thus is not marked as optional